BREAKING(build): Enhance build automation, Update gulp-sass

plutonik-a opened the pull request:
**Introduces a dependency to Node version manager [nvm](https://github.com/nvm-sh/nvm)**

1. Update gulp-sass:
    * delete deprecated node-sass dependency in gulpfile
    * instead, import sass compiler "sass"
    * update gulpfile accordingly: See https://www.npmjs.com/package/gulp-sass and https://github.com/dlmanning/gulp-sass/wiki/Common-Issues-and-Their-Fixes, and
https://github.com/dlmanning/gulp-sass/wiki/Update-to-the-latest-Gulp-Sass
1. Enhance build automation:
    * automate creating a build.properties.local file - run `ant prepare`.
    * introduce a dependency to nvm as a node version managing tool in
order to resolve a development dependency tied to a specific node
version
    * specify the current node version in a .nvmrc lockfile - run `ant node` to internally call `nvm use` to install the currently specified node version of the project
    * update readme with new installation and build instructions
    * the full installation / build command (see below) is only required in
case a node version has changed in the project
    * otherwise the default task `ant` can be used as usual

    The new complete command for building a xar file from scratch is:
`ant prepare && ant node && ant`

joewiz commented:
@plutonik-a When I run `ant prepare` as directed in the PR text I get the following output, which suggests that the `${node.PATH}` property is not defined:

```
joe@choskimac-iii hsg-shell % ant prepare   
Buildfile: /Users/joe/workspace/hsg-project/repos/hsg-shell/build.xml
     [echo] ------------------------------------------------------------
     [echo] Prepare local build properties and node version
     [echo] ------------------------------------------------------------

prepare:
     [echo] Calling ant target 'copy-paths-to-local-build-properties'
     [echo] ------------------------------------------------------------
     [echo] Prepare local build properties and node version
     [echo] ------------------------------------------------------------

get-binary-paths:
     [echo] Call target get-binary-paths
     [echo] Path to gulp binary: /Users/joe/.nvm/versions/node/v10.16.0/bin/gulp
     [echo] Path to node binary: /Users/joe/.nvm/versions/node/v10.16.0/bin/node
     [echo] Path to npm binary: /Users/joe/.nvm/versions/node/v10.16.0/bin/npm

copy-paths-to-local-build-properties:
     [echo] Copy binary paths to build.properties.local.xml
     [copy] Copying 1 file to /Users/joe/workspace/hsg-project/repos/hsg-shell
     [echo] Copied /Users/joe/.nvm/versions/node/v10.16.0/bin/npm, /Users/joe/.nvm/versions/node/v10.16.0/bin/gulp, and ${node.PATH} to local.build.properties

BUILD SUCCESSFUL
Total time: 0 seconds
```

Looking in the `build.properties.xml` template file, I see filter tokens for `@npm` and `@gulp` but not `@nodePath`. Could this be a clue?

In the update to the README.md, the Update section says to execute the following command to create an up-to-date build package: `git pull && ant prepare && ant node && ant`. But when I run this command there is an error on `git pull`:

```
joe@choskimac-iii hsg-shell % git pull && ant prepare && ant node && ant
error: Pulling is not possible because you have unmerged files.
hint: Fix them up in the work tree, and then use 'git add/rm <file>'
hint: as appropriate to mark resolution and make a commit.
fatal: Exiting because of an unresolved conflict.
```

Checking what's unmerged...

```
joe@choskimac-iii hsg-shell % git status
On branch feat/enhance-build-automation
Your branch is up to date with 'origin/feat/enhance-build-automation'.

Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
	modified:   resources/images/Office-of-the-Historian-logo_500x168.jpg

Unmerged paths:
  (use "git restore --staged <file>..." to unstage)
  (use "git add <file>..." to mark resolution)
	both modified:   package-lock.json

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   resources/images/Office-of-the-Historian-logo_500x168.jpg
```

... so this step is not working. Could you please advise on how we should approach this?

Lastly, it seems that `ant prepare` and `ant node` execute very quickly, once run the first time. Is there a reason we should not create a new target (say, `all`) and add these to the default target? I ask because in hsg-project, the main build file is quite blunt: when building and deploying each package in the `repos` subdirectory, it simply calls `ant` in each subdirectory. There is no provision for a separate, per-repo setup step in hsg-project. Would there be a strong disadvantage to adding this?

```xml
<target name="all" depends="prepare,node,xar"/>
```

... and switching the root to `<project default="all">`?

plutonik-a commented:
Hi @joewiz 
1. Re: "${node.PATH} property is not defined":  
Sure, in the ant task prepare, the variable ${node.PATH} is not necessary and should be removed from the filterlist, because at this point, I just copy the existing and fixed value to the new local build properties file.
2. Re: Merge-conflict with image: 
This happens due to the default image optimization task in gulp. I have extracted this task, as it doesn’t need to run per default on each build. The only image in the hsg-shell application is the logo, and it has already been optimized in the repo. 
3. Re: Merge ant tasks into one call:
Your suggestion `<target name="all" depends="prepare,node,xar"/>` was my first solution, but I ran into race conditions when calling the tasks as a sequence. For some reason, calling the tasks as separate commands solved this problem, even if the processing time is identical in the end. There must be some underlying bug, I guess. Have you tried your solution and it succeeded? Maybe this is just a problem on my system.

joewiz commented:
@plutonik-a I think this is working now as expected. Below are my notes. 

## ensure a clean environment

i performed these steps to ensure nothing was left over from old versions of node, npm, etc.:

- uninstalled node(s) installed with homebrew, via `brew list` and then `brew uninstall` on the installed versions (`node` and `node@10`)
- deleted global npm packages fully, via `rm -rf /usr/local/lib/node_modules`
- uninstalled nvm fully, via `rm -rf ~/.nvm`
- deleted any existing node/npm entries in `~/.zshrc`
- checked out this PR branch of hsg-shell and ran `git clean -xdf` to remove any artifacts

## install nvm

i followed the directions at https://github.com/nvm-sh/nvm#install--update-script:

```
% curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.39.1/install.sh | bash  
 % Total % Received % Xferd Average Speed Time Time Time Current  
 Dload Upload Total Spent Left Speed  
100 15037 100 15037 0 0 35064 0 --:--:-- --:--:-- --:--:-- 35548  
=> Downloading nvm from git to '/Users/joe/.nvm'  
=> Cloning into '/Users/joe/.nvm'...  
remote: Enumerating objects: 354, done.  
remote: Counting objects: 100% (354/354), done.  
remote: Compressing objects: 100% (302/302), done.  
remote: Total 354 (delta 40), reused 156 (delta 27), pack-reused 0  
Receiving objects: 100% (354/354), 206.94 KiB | 790.00 KiB/s, done.  
Resolving deltas: 100% (40/40), done.  
* (HEAD detached at FETCH_HEAD)  
 master  
=> Compressing and cleaning up git repository  
  
=> nvm source string already in /Users/joe/.zshrc  
=> bash_completion source string already in /Users/joe/.zshrc  
env: node: No such file or directory  
=> Close and reopen your terminal to start using nvm or run the following to use it now:  
  
export NVM_DIR="$HOME/.nvm"  
[ -s "$NVM_DIR/nvm.sh" ] && \. "$NVM_DIR/nvm.sh" # This loads nvm  
[ -s "$NVM_DIR/bash_completion" ] && \. "$NVM_DIR/bash_completion" # This loads nvm bash_completion
```

then open new shell and confirm the NVM_DIR entries are in `~/.zshrc`:

```
% cat ~/.zshrc
#
# other entries trimmed for brevity
#

export NVM_DIR="$HOME/.nvm"
[ -s "$NVM_DIR/nvm.sh" ] && \. "$NVM_DIR/nvm.sh"  # This loads nvm
[ -s "$NVM_DIR/bash_completion" ] && \. "$NVM_DIR/bash_completion"  # This loads nvm bash_completion
```

## install gulp and bower

from the README, ran this command

```
% npm install -g gulp bower
zsh: command not found: npm
```

conclusion: looks like we need to install npm. 

## install npm

checking `nvm -h` to see how to install npm, the best approach seems to be calling `nvm install` (which installs the version of node defined in `.nvmrc`).

```
% nvm install   
Found '/Users/joe/workspace/hsg-project/repos/hsg-shell/.nvmrc' with version <10.16.0>
Downloading and installing node v10.16.0...
Downloading https://nodejs.org/dist/v10.16.0/node-v10.16.0-darwin-x64.tar.xz...
######################################################################################################################################################################################################################################################### 100.0%
Computing checksum with shasum -a 256
Checksums matched!
Now using node v10.16.0 (npm v6.9.0)
Creating default alias: default -> 10.16.0 (-> v10.16.0)
```

Note: I first tried the `--lastest-npm` flag (defined as "After installing, attempt to upgrade to the latest working npm on the given node version"). This updated to v7.24.2, saying "`npm` `v7.x` is the last version that works on `node` `v13`, `v15`, below `v12.13`, or `v14.0` - `v14.15`"—but I decided to back out of this - since I wasn't sure if using a newer npm is ok... It also instructed me to update to a newer version: "New major version of npm available! 6.9.0 → 8.4.0". But given the info about v7.x being the only version compatible with node under v12.13, I trust we shouldn't upgrade to v8?

## retry gulp and bower installation

this time it seemed to work!

```
% npm install -g gulp bower
npm WARN deprecated chokidar@2.1.8: Chokidar 2 will break on node v14+. Upgrade to chokidar 3 with 15x less dependencies.
npm WARN deprecated resolve-url@0.2.1: https://github.com/lydell/resolve-url#deprecated
npm WARN deprecated urix@0.1.0: Please see https://github.com/lydell/urix#deprecated
npm WARN deprecated fsevents@1.2.13: fsevents 1 will break on node v14+ and could be using insecure binaries. Upgrade to fsevents 2.
/Users/joe/.nvm/versions/node/v10.16.0/bin/bower -> /Users/joe/.nvm/versions/node/v10.16.0/lib/node_modules/bower/bin/bower
/Users/joe/.nvm/versions/node/v10.16.0/bin/gulp -> /Users/joe/.nvm/versions/node/v10.16.0/lib/node_modules/gulp/bin/gulp.js

> fsevents@1.2.13 install /Users/joe/.nvm/versions/node/v10.16.0/lib/node_modules/gulp/node_modules/fsevents
> node install.js

  SOLINK_MODULE(target) Release/.node
  CXX(target) Release/obj.target/fse/fsevents.o
  SOLINK_MODULE(target) Release/fse.node
+ gulp@4.0.2
+ bower@1.8.13
added 329 packages from 228 contributors in 16.505s
```

## tried unifying the 3-target ant command into 1 "all" which depends on the 3

i thought I'd see if I also encountered the "race conditions" you did, and indeed, i did. 

the target:

```xml
<target name="all" depends="prepare,node,xar"/>
```

... and i changed the `project/@default` to `all`.

the result:

```
% ant
Buildfile: /Users/joe/workspace/hsg-project/repos/hsg-shell/build.xml
     [echo] ------------------------------------------------------------
     [echo] Prepare local build properties and node version
     [echo] ------------------------------------------------------------

prepare:
     [echo] Calling ant target 'copy-paths-to-local-build-properties'
     [echo] ------------------------------------------------------------
     [echo] Prepare local build properties and node version
     [echo] ------------------------------------------------------------

get-binary-paths:
     [echo] Call target get-binary-paths
     [echo] Path to gulp binary: /Users/joe/.nvm/versions/node/v10.16.0/bin/gulp
     [echo] Path to node binary: /Users/joe/.nvm/versions/node/v10.16.0/bin/node
     [echo] Path to npm binary: /Users/joe/.nvm/versions/node/v10.16.0/bin/npm

copy-paths-to-local-build-properties:
     [echo] Copy binary paths to build.properties.local.xml
     [copy] Copying 1 file to /Users/joe/workspace/hsg-project/repos/hsg-shell
     [echo] Copied /Users/joe/.nvm/versions/node/v10.16.0/bin/npm and /Users/joe/.nvm/versions/node/v10.16.0/bin/gulp to local.build.properties

node:
     [echo] Calling ant target 'install-project-specific-node-version'
     [echo] ------------------------------------------------------------
     [echo] Prepare local build properties and node version
     [echo] ------------------------------------------------------------

install-project-specific-node-version:
     [echo] Install node version from nvmrc file...
     [echo] 
     [echo] > hsg-shell@1.0.0 nvm /Users/joe/workspace/hsg-project/repos/hsg-shell
     [echo] > source ~/.nvm/nvm.sh && nvm use
     [echo] 
     [echo] Found '/Users/joe/workspace/hsg-project/repos/hsg-shell/.nvmrc' with version <10.16.0>
     [echo] Now using node v10.16.0 (npm v6.9.0)

clean:
     [echo] Calling gulp clean...

BUILD FAILED
/Users/joe/workspace/hsg-project/repos/hsg-shell/build.xml:28: no executable specified
```

this failure is understandable, since ant can't read the properties it writes to the local-properties-build file until it is run again.

so it seems that at least for first time setup, the 3-target ant command must be run.

## run 3-target ant command

first, clean up stuff left from previous attempt; then run the 3-target ant command.

```
% git clean -xdf
Removing build.properties.local.xml
% git checkout .
Updated 1 path from the index
% ant prepare && ant node && ant   
Buildfile: /Users/joe/workspace/hsg-project/repos/hsg-shell/build.xml
     [echo] ------------------------------------------------------------
     [echo] Prepare local build properties and node version
     [echo] ------------------------------------------------------------

prepare:
     [echo] Calling ant target 'copy-paths-to-local-build-properties'
     [echo] ------------------------------------------------------------
     [echo] Prepare local build properties and node version
     [echo] ------------------------------------------------------------

get-binary-paths:
     [echo] Call target get-binary-paths
     [echo] Path to gulp binary: /Users/joe/.nvm/versions/node/v10.16.0/bin/gulp
     [echo] Path to node binary: /Users/joe/.nvm/versions/node/v10.16.0/bin/node
     [echo] Path to npm binary: /Users/joe/.nvm/versions/node/v10.16.0/bin/npm

copy-paths-to-local-build-properties:
     [echo] Copy binary paths to build.properties.local.xml
     [copy] Copying 1 file to /Users/joe/workspace/hsg-project/repos/hsg-shell
     [echo] Copied /Users/joe/.nvm/versions/node/v10.16.0/bin/npm and /Users/joe/.nvm/versions/node/v10.16.0/bin/gulp to local.build.properties

BUILD SUCCESSFUL
Total time: 0 seconds
Buildfile: /Users/joe/workspace/hsg-project/repos/hsg-shell/build.xml
     [echo] ------------------------------------------------------------
     [echo] Prepare local build properties and node version
     [echo] ------------------------------------------------------------

node:
     [echo] Calling ant target 'install-project-specific-node-version'
     [echo] ------------------------------------------------------------
     [echo] Prepare local build properties and node version
     [echo] ------------------------------------------------------------

install-project-specific-node-version:
     [echo] Install node version from nvmrc file...
     [echo] 
     [echo] > hsg-shell@1.0.0 nvm /Users/joe/workspace/hsg-project/repos/hsg-shell
     [echo] > source ~/.nvm/nvm.sh && nvm use
     [echo] 
     [echo] Found '/Users/joe/workspace/hsg-project/repos/hsg-shell/.nvmrc' with version <10.16.0>
     [echo] Now using node v10.16.0 (npm v6.9.0)

BUILD SUCCESSFUL
Total time: 0 seconds
Buildfile: /Users/joe/workspace/hsg-project/repos/hsg-shell/build.xml
     [echo] ------------------------------------------------------------
     [echo] Prepare local build properties and node version
     [echo] ------------------------------------------------------------

clean:
     [echo] Calling gulp clean...
     [exec] Result: 1
     [echo] [20:39:01] Local modules not found in ~/workspace/hsg-project/repos/hsg-shell
     [echo] [20:39:01] Try running: npm install
     [echo] Deleting xar files...

run-npm:
     [echo] Calling npm start... (Production? false)
     [echo] 
     [echo] > hsg-shell@1.0.0 start /Users/joe/workspace/hsg-project/repos/hsg-shell
     [echo] > npm install --quiet && bower install --quiet && gulp
     [echo] 
     [echo] 
     [echo] > fibers_node_v8@3.1.5 preinstall /Users/joe/workspace/hsg-project/repos/hsg-shell/node_modules/fibers_node_v8
     [echo] > node preinstall.js
     [echo] 
     [echo] 
     [echo] > fibers@4.0.3 install /Users/joe/workspace/hsg-project/repos/hsg-shell/node_modules/fibers
     [echo] > node build.js || nodejs build.js
     [echo] 
     [echo] `darwin-x64-64` exists; testing
     [echo] Binary is fine; exiting
     [echo] 
     [echo] > fibers_node_v8@3.1.5 install /Users/joe/workspace/hsg-project/repos/hsg-shell/node_modules/fibers_node_v8
     [echo] > node build.js
     [echo] 
     [echo] ignore install
     [echo] 
     [echo] > fsevents@1.2.13 install /Users/joe/workspace/hsg-project/repos/hsg-shell/node_modules/glob-watcher/node_modules/fsevents
     [echo] > node install.js
     [echo] 
     [echo]   SOLINK_MODULE(target) Release/.node
     [echo]   CXX(target) Release/obj.target/fse/fsevents.o
     [echo]   SOLINK_MODULE(target) Release/fse.node
     [echo] 
     [echo] > chromedriver@91.0.1 install /Users/joe/workspace/hsg-project/repos/hsg-shell/node_modules/chromedriver
     [echo] > node install.js
     [echo] 
     [echo] (node:60120) ExperimentalWarning: The fs.promises API is experimental
     [echo] Your Chrome version is 97.0.4692.99
     [echo] Finding Chromedriver version.
     [echo] Chromedriver version is 97.0.4692.71.
     [echo] Compatible ChromeDriver version is 97.0.4692.71
     [echo] ChromeDriver binary exists. Validating...
     [echo] ChromeDriver is already available at '/var/folders/63/20jpnkvx09s7q6y9qmgdkj3h0000gn/T/97.0.4692.71/chromedriver/chromedriver'.
     [echo] Copying to target path /Users/joe/workspace/hsg-project/repos/hsg-shell/node_modules/chromedriver/lib/chromedriver
     [echo] Fixing file permissions.
     [echo] Done. ChromeDriver binary available at /Users/joe/workspace/hsg-project/repos/hsg-shell/node_modules/chromedriver/lib/chromedriver/chromedriver
     [echo] 
     [echo] > gifsicle@5.2.1 postinstall /Users/joe/workspace/hsg-project/repos/hsg-shell/node_modules/gifsicle
     [echo] > node lib/install.js
     [echo] 
     [echo] gifsicle pre-build test passed successfully
     [echo] 
     [echo] > mozjpeg@6.0.1 postinstall /Users/joe/workspace/hsg-project/repos/hsg-shell/node_modules/mozjpeg
     [echo] > node lib/install.js
     [echo] 
     [echo] 
     [echo] > optipng-bin@6.0.0 postinstall /Users/joe/workspace/hsg-project/repos/hsg-shell/node_modules/optipng-bin
     [echo] > node lib/install.js
     [echo] 
     [echo]   ✔ mozjpeg pre-build test passed successfully
     [echo]   ✔ optipng pre-build test passed successfully
     [echo] added 1186 packages from 622 contributors and audited 1189 packages in 16.267s
     [echo] found 26 vulnerabilities (20 moderate, 6 high)
     [echo]   run `npm audit fix` to fix them, or `npm audit` for details
     [echo] Production? false
     [echo] read from localConnectionOptions { host: 'localhost',
     [echo]   port: '8080',
     [echo]   path: '/exist/xmlrpc',
     [echo]   basic_auth: { user: 'admin', pass: '' } }
     [echo] [20:39:21] Using gulpfile ~/workspace/hsg-project/repos/hsg-shell/gulpfile.js
     [echo] [20:39:21] Starting 'default'...
     [echo] [20:39:21] Starting 'build'...
     [echo] [20:39:21] Starting 'clean'...
     [echo] [20:39:21] Finished 'clean' after 12 ms
     [echo] [20:39:21] Starting 'fonts:copy'...
     [echo] [20:39:21] Starting 'styles:concat'...
     [echo] [20:39:21] Starting 'templates:build'...
     [echo] [20:39:21] Starting 'scripts:concat'...
     [echo] [20:39:21] Starting '<anonymous>'...
     [echo] [20:39:21] Starting 'styles:build'...
     [echo] [20:39:21] Starting 'scripts:build'...
     [echo] [20:39:21] Starting '<anonymous>'...
     [echo] DEPRECATION WARNING: Using / for division is deprecated and will be removed in Dart Sass 2.0.0.
     [echo] 
     [echo] Recommendation: math.div($grid-gutter-width, 2)
     [echo] 
     [echo] More info and automated migrator: https://sass-lang.com/d/slash-div
     [echo] 
     [echo]     ╷
     [echo] 368 │ $navbar-padding-horizontal:        floor(($grid-gutter-width / 2)) !default;
     [echo]     │                                           ^^^^^^^^^^^^^^^^^^^^^^
     [echo]     ╵
     [echo]     bower_components/bootstrap-sass/assets/stylesheets/bootstrap/_variables.scss 368:43  @import
     [echo]     bower_components/bootstrap-sass/assets/stylesheets/_bootstrap.scss 8:9               @import
     [echo]     app/scss/main.scss 7:9                                                               root stylesheet
     [echo] 
     [echo] DEPRECATION WARNING: Using / for division is deprecated and will be removed in Dart Sass 2.0.0.
     [echo] 
     [echo] Recommendation: math.div($navbar-height - $line-height-computed, 2)
     [echo] 
     [echo] More info and automated migrator: https://sass-lang.com/d/slash-div
     [echo] 
     [echo]     ╷
     [echo] 369 │ $navbar-padding-vertical:          (($navbar-height - $line-height-computed) / 2) !default;
     [echo]     │                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
     [echo]     ╵
     [echo]     bower_components/bootstrap-sass/assets/stylesheets/bootstrap/_variables.scss 369:37  @import
     [echo]     bower_components/bootstrap-sass/assets/stylesheets/_bootstrap.scss 8:9               @import
     [echo]     app/scss/main.scss 7:9                                                               root stylesheet
     [echo] 
     [echo] DEPRECATION WARNING: Using / for division is deprecated and will be removed in Dart Sass 2.0.0.
     [echo] 
     [echo] Recommendation: math.div($line-height-computed, 2)
     [echo] 
     [echo] More info and automated migrator: https://sass-lang.com/d/slash-div
     [echo] 
     [echo]    ╷
     [echo] 28 │   margin-bottom: ($line-height-computed / 2);
     [echo]    │                   ^^^^^^^^^^^^^^^^^^^^^^^^^
     [echo]    ╵
     [echo]     bower_components/bootstrap-sass/assets/stylesheets/bootstrap/_type.scss 28:19  @import
     [echo]     bower_components/bootstrap-sass/assets/stylesheets/_bootstrap.scss 18:9        @import
     [echo]     app/scss/main.scss 7:9                                                         root stylesheet
     [echo] 
     [echo] DEPRECATION WARNING: Using / for division is deprecated and will be removed in Dart Sass 2.0.0.
     [echo] 
     [echo] Recommendation: math.div($line-height-computed, 2)
     [echo] 
     [echo] More info and automated migrator: https://sass-lang.com/d/slash-div
     [echo] 
     [echo]    ╷
     [echo] 38 │   margin-top: ($line-height-computed / 2);
     [echo]    │                ^^^^^^^^^^^^^^^^^^^^^^^^^
     [echo]    ╵
     [echo]     bower_components/bootstrap-sass/assets/stylesheets/bootstrap/_type.scss 38:16  @import
     [echo]     bower_components/bootstrap-sass/assets/stylesheets/_bootstrap.scss 18:9        @import
     [echo]     app/scss/main.scss 7:9                                                         root stylesheet
     [echo] 
     [echo] DEPRECATION WARNING: Using / for division is deprecated and will be removed in Dart Sass 2.0.0.
     [echo] 
     [echo] Recommendation: math.div($line-height-computed, 2)
     [echo] 
     [echo] More info and automated migrator: https://sass-lang.com/d/slash-div
     [echo] 
     [echo]    ╷
     [echo] 39 │   margin-bottom: ($line-height-computed / 2);
     [echo]    │                   ^^^^^^^^^^^^^^^^^^^^^^^^^
     [echo]    ╵
     [echo]     bower_components/bootstrap-sass/assets/stylesheets/bootstrap/_type.scss 39:19  @import
     [echo]     bower_components/bootstrap-sass/assets/stylesheets/_bootstrap.scss 18:9        @import
     [echo]     app/scss/main.scss 7:9                                                         root stylesheet
     [echo] 
     [echo] WARNING: 34 repetitive deprecation warnings omitted.
     [echo] 
     [echo] [20:39:23] Finished 'templates:build' after 2.56 s
     [echo] [20:39:23] Finished '<anonymous>' after 2.55 s
     [echo] [20:39:23] Finished 'styles:build' after 2.56 s
     [echo] [20:39:23] Starting '<anonymous>'...
     [echo] [20:39:23] Finished 'scripts:build' after 2.56 s
     [echo] [20:39:23] Starting '<anonymous>'...
     [echo] [20:39:23] Finished '<anonymous>' after 5.25 ms
     [echo] [20:39:23] Finished 'scripts:concat' after 2.58 s
     [echo] [20:39:23] Finished '<anonymous>' after 21 ms
     [echo] [20:39:23] Finished 'styles:concat' after 2.58 s
     [echo] [20:39:23] Finished '<anonymous>' after 2.58 s
     [echo] [20:39:23] Finished 'fonts:copy' after 2.58 s
     [echo] [20:39:23] Finished 'build' after 2.6 s
     [echo] [20:39:23] Finished 'default' after 2.6 s

git.revision:
     [echo] Git repo: a12c80a38c61b9a4dec29d497bf39c0c4b49f232
     [echo] Git time: 1638201711

xar:
     [echo] Creating build folder...
    [mkdir] Created dir: /Users/joe/workspace/hsg-project/repos/hsg-shell/build
     [copy] Copying 2 files to /Users/joe/workspace/hsg-project/repos/hsg-shell
     [copy] Copying /Users/joe/workspace/hsg-project/repos/hsg-shell/expath-pkg.xml.tmpl to /Users/joe/workspace/hsg-project/repos/hsg-shell/expath-pkg.xml
     [copy] Copying /Users/joe/workspace/hsg-project/repos/hsg-shell/repo.xml.tmpl to /Users/joe/workspace/hsg-project/repos/hsg-shell/repo.xml
     [echo] ------------------------------------------------------------
     [echo] Creating xar file...
     [echo] ------------------------------------------------------------
      [zip] Building zip: /Users/joe/workspace/hsg-project/repos/hsg-shell/build/hsg-shell-2.3.0.xar

BUILD SUCCESSFUL
Total time: 23 seconds
```

## installation of xar 

installation of the resulting works whether the xar is generated via the command above or via oXygen's "deploy current file's repository to localhost"

## questions

- is it expected that besides installing nvm we need to explicitly install npm too? 
- should we stick with npm 6.9.0 or use `--install-latest-npm` to upgrade from 6.9.0 to 7.24.2?
- could you confirm that we should not follow npm's advice ("New major version of npm available! 6.9.0 → 8.4.0", `npm install -g npm`)? (This same command is also recommended in the README.)
